grafana-loki,promtail: 2.9.6 -> 3.0.0

r-ryantm opened the pull request:
Automatic update generated by [nixpkgs-update](https://github.com/ryantm/nixpkgs-update) tools. This update was made based on information from https://github.com/grafana/loki/releases.


meta.description for grafana-loki is: Like Prometheus, but for logs

meta.homepage for grafana-loki is: https://grafana.com/oss/loki/

meta.changelog for grafana-loki is: https://github.com/grafana/loki/releases/tag/v3.0.0


###### Updates performed
- Golang update

###### To inspect upstream changes

- [Release on GitHub](https://github.com/grafana/loki/releases/tag/v3.0.0)

- [Compare changes on GitHub](https://github.com/grafana/loki/compare/v2.9.6...v3.0.0)

###### Impact

<b>Checks done</b>

---

- built on NixOS
- The tests defined in `passthru.tests`, if any, passed
- found 3.0.0 with grep in /nix/store/0mhj4f899qznjag5m1fvkgya40lrabzj-grafana-loki-3.0.0
- found 3.0.0 in filename of file in /nix/store/0mhj4f899qznjag5m1fvkgya40lrabzj-grafana-loki-3.0.0

---

<details>
<summary>
<b>Rebuild report</b> (if merged into master) (click to expand)
</summary>

```
2 total rebuild path(s)

2 package rebuild(s)

First fifty rebuilds by attrpath
grafana-loki
promtail
```

</details>

<details>
<summary>
<b>Instructions to test this update</b> (click to expand)
</summary>

---

Either **download from Cachix**:
```
nix-store -r /nix/store/0mhj4f899qznjag5m1fvkgya40lrabzj-grafana-loki-3.0.0 \
  --option binary-caches 'https://cache.nixos.org/ https://nix-community.cachix.org/' \
  --option trusted-public-keys '
  nix-community.cachix.org-1:mB9FSh9qf2dCimDSUo8Zy7bkq5CX+/rkCWyvRCYg3Fs=
  cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=
  '
```
(The Cachix cache is only trusted for this store-path realization.)
For the Cachix download to work, your user must be in the `trusted-users` list or you can use `sudo` since root is effectively trusted.

Or, **build yourself**:
```
nix-build -A grafana-loki https://github.com/r-ryantm/nixpkgs/archive/21171e93e9aefd1baceae89ef6faa76a42d90652.tar.gz
```
Or:
```
nix build github:r-ryantm/nixpkgs/21171e93e9aefd1baceae89ef6faa76a42d90652#grafana-loki
```

After you've downloaded or built it, look at the files and if there are any, run the binaries:
```
ls -la /nix/store/0mhj4f899qznjag5m1fvkgya40lrabzj-grafana-loki-3.0.0
ls -la /nix/store/0mhj4f899qznjag5m1fvkgya40lrabzj-grafana-loki-3.0.0/bin
```

---

</details>
<br/>



### Pre-merge build results

We have automatically built all packages that will get rebuilt due to
this change.

This gives evidence on whether the upgrade will break dependent packages.
Note sometimes packages show up as _failed to build_ independent of the
change, simply because they are already broken on the target branch.

Result of `nixpkgs-review --extra-nixpkgs-config '{ allowInsecurePredicate = x: true; }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>grafana-loki</li>
    <li>promtail</li>
  </ul>
</details>

---

###### Maintainer pings

cc @WilliButz @globin @mmahut @emilylange @dasJ for [testing](https://github.com/ryantm/nixpkgs-update/blob/master/doc/nixpkgs-maintainer-faq.md#r-ryantm-opened-a-pr-for-my-package-what-do-i-do).

emilylange reviewed (changes requested):
Upstream made changes to the ldflag prefix responsible for versioning.

See https://github.com/grafana/loki/commit/b4f7181c7aa9484e66976e8a933111a9b85ea8c2

The version details we provide are no longer getting picked up.

```bash
❯ git describe --abbrev=40
24.05-pre-56881-g21171e93e9aefd1baceae89ef6faa76a42d90652

❯ result/bin/loki --version
loki, version  (branch: , revision: unknown)
  build user:       
  build date:       
  go version:       go1.22.1
  platform:         linux/amd64
  tags:             promtail_journal_enabled
```

Do not merge as is.

I'll redo the commit and will also add a version test.

emilylange committed:
grafana-loki,promtail: 2.9.6 -> 3.0.0

Starting with v3.0.0, the ldflag prefix now also includes `v3`.

It's unclear if this will change with each major release.
This is something we should keep an eye on.

The commit right after this will add a simple version test for this
based on `testers.testVersion`.

https://github.com/grafana/loki/releases/tag/v3.0.0

changelog: https://github.com/grafana/loki/blob/c9142722684ad6aca2bb09ae5487d5a2af087d0f/CHANGELOG.md

diff (unusable, 5000+ edits because upstream vendors go deps): https://github.com/grafana/loki/compare/v2.9.6...v3.0.0

emilylange reviewed (approved):
done^^

(would be cool if someone else could review my changes)

wegank commented on pkgs/servers/monitoring/loki/default.nix:
```diff
@@ -38,9 +40,15 @@ buildGoModule rec {
+  passthru.tests = {
+    inherit (nixosTests) loki;
+    version = testers.testVersion {
+      command = "${grafana-loki}/bin/loki --version";
```
```suggestion
      command = "loki --version";
```

Does this work?

emilylange committed:
grafana-loki: add simple version test

v3.0.0 changed the ldflag prefix, which we rely on to pass and embed
version information.

This is likely to happen again, and such version test is incredibly
inexpensive.
